sdks/java/core/src/test/java/org/apache/beam/sdk/runners/TransformHierarchyTest.java

```diff
@@ -492,4 +495,87 @@ public void visitPrimitiveTransform(Node node) {
     assertThat(visitedPrimitiveNodes, containsInAnyOrder(upstreamNode, replacementParNode));
     assertThat(visitedValues, Matchers.<PValue>containsInAnyOrder(upstream, output));
   }
+
+  @Test
+  public void visitIsTopologicallyOrdered() {
```

Comment:
if I were changing this code, I'd prefer to have a set of small tests structured like "doesn't revisit composite when arriving at child".